code/campaign/src/main/java/com/adobe/marketing/mobile/campaign/CampaignExtension.java

```diff
@@ -240,35 +244,50 @@ public boolean readyForEvent(final @NonNull Event event) {
 
     /**
      * Processes all events dispatched to the {@code EventHub} to determine if any rules are matched.
-     * <p>
-     * If a rule is triggered then an appropriate {@link CampaignMessage} object is instantiated and the message is shown.
      *
      * @param event incoming {@link Event} object to be processed
      */
     void handleWildcardEvents(final Event event) {
-        List<LaunchRule> triggeredRules = campaignRulesEngine.process(event);
-        final List<RuleConsequence> consequences = new ArrayList<>();
-
-        if (triggeredRules == null || triggeredRules.isEmpty()) {
+        if (event == null) {
+            Log.trace(CampaignConstants.LOG_TAG, SELF_TAG, "handleWildcardEvents - ignoring null event received.");
             return;
         }
 
-        for (final LaunchRule rule : triggeredRules) {
-            consequences.addAll(rule.getConsequenceList());
+        campaignRulesEngine.processEvent(event);
+    }
+
+    /**
+     * Handles Rule Engine Response Content events which are dispatched when a event matches a rule in the Campaign {@link LaunchRulesEngine}.
+     * This handler will attempt to  show a {@link CampaignMessage} created from the triggered rule consequence.
+     *
+     * @param event incoming {@link Event} object to be processed
+     */
+    void handleRuleEngineResponseEvents(final Event event) {
+        final Map<String, Object> consequenceMap = DataReader.optTypedMap(Object.class, event.getEventData(), CampaignConstants.EventDataKeys.RuleEngine.CONSEQUENCE_TRIGGERED, null);
+
+        if (MapUtils.isNullOrEmpty(consequenceMap)) {
+            Log.trace(CampaignConstants.LOG_TAG, SELF_TAG, "handleRulesResponseEvents - null or empty consequences found. Will not handle rules response event.");
+            return;
         }
 
-        if (consequences.isEmpty()) {
+        final String id = DataReader.optString(consequenceMap, CampaignConstants.EventDataKeys.RuleEngine.MESSAGE_CONSEQUENCE_ID, "");
+        final String type = DataReader.optString(consequenceMap, CampaignConstants.EventDataKeys.RuleEngine.MESSAGE_CONSEQUENCE_TYPE, "");
+        final Map<String, Object> detail = DataReader.optTypedMap(Object.class, consequenceMap, CampaignConstants.EventDataKeys.RuleEngine.MESSAGE_CONSEQUENCE_DETAIL, null);
+
+        // detail is required
+        if (MapUtils.isNullOrEmpty(detail)) {
+            Log.trace(CampaignConstants.LOG_TAG, SELF_TAG, "handleRulesResponseEvents - null or empty consequence details found. Will not handle rules response event.");
             return;
         }
 
         try {
-            final CampaignMessage triggeredMessage = CampaignMessage.createMessageObject(this, consequences.get(0));
+            final CampaignMessage triggeredMessage = CampaignMessage.createMessageObject(this, new RuleConsequence(id, type, detail));
 
             if (triggeredMessage != null) {
```

Comment:
should we log in case the `CampaignMessage` fails to initialize?

Reply:
yup makes sense, will update this